gradle-plugin/src/gradleTest/java/com/squareup/anvil/plugin/IncrementalTest.kt

```diff
@@ -0,0 +1,545 @@
+import io.kotest.matchers.file.shouldNotExist
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.string.shouldContain
+import org.gradle.testkit.runner.TaskOutcome
+import org.junit.jupiter.api.DynamicNode
+import org.junit.jupiter.api.TestFactory
+import java.util.stream.Stream
+
+class IncrementalTest : BaseGradleTest() {
+
+  @TestFactory
+  fun `compileKotlin should execute again if anvil-generated output is deleted`() = testFactory {
+
+    rootProject {
+      dir("src/main/java") {
+        injectClass()
+        kotlinFile(
+          "com/squareup/test/OtherClass.kt",
+          """
+          package com.squareup.test
+      
+          import javax.inject.Inject
+          
+          class OtherClass @Inject constructor()
+          """.trimIndent(),
+        )
+      }
+      gradlePropertiesFile(
+        """
+        com.squareup.anvil.trackSourceFiles=true
+        """.trimIndent(),
+      )
+    }
+
+    shouldSucceed("compileJava")
+
+    val injectClassFactory = rootAnvilMainGenerated
+      .resolve("com/squareup/test/InjectClass_Factory.kt")
+
+    injectClassFactory.shouldExist()
+    injectClassFactory.delete()
+    injectClassFactory.shouldNotExist()
+
+    shouldSucceed("compileJava") {
+      task(":compileKotlin")?.outcome shouldBe TaskOutcome.SUCCESS
+    }
+
+    injectClassFactory.shouldExist()
+  }
+
+  @TestFactory
+  fun `a generated factory is deleted if the source file is deleted`() = testFactory {
+
+    val injectClassPath = "com/squareup/test/InjectClass.kt"
+
+    rootProject {
+      dir("src/main/java") {
+
+        kotlinFile(
+          injectClassPath,
+          """
+          package com.squareup.test
+      
+          import javax.inject.Inject
+          
+          class InjectClass @Inject constructor()
+          """.trimIndent(),
+        )
+
+        kotlinFile(
+          "com.squareup/test/OtherClass.kt",
+          """
+          package com.squareup.test
+      
+          import javax.inject.Inject
+          
+          class OtherClass @Inject constructor()
+          """.trimIndent(),
+        )
+      }
+      gradlePropertiesFile(
+        """
+        com.squareup.anvil.trackSourceFiles=true
+        """.trimIndent(),
+      )
+    }
+
+    shouldSucceed("compileJava")
+
+    rootAnvilMainGenerated.injectClassFactory.shouldExist()
+    rootAnvilMainGenerated.resolve("com/squareup/test/OtherClass_Factory.kt").shouldExist()
+
+    workingDir.resolve("src/main/java")
+      .resolve(injectClassPath)
+      .delete()
+
+    shouldSucceed("compileJava") {
+      task(":compileKotlin")?.outcome shouldBe TaskOutcome.SUCCESS
+    }
+
+    rootAnvilMainGenerated.injectClassFactory.shouldNotExist()
+    rootAnvilMainGenerated.resolve("com/squareup/test/OtherClass_Factory.kt").shouldExist()
+  }
+
+  @TestFactory
+  fun `a generated factory is deleted if the @Inject annotation is removed`() =
+    withTrackSourceFiles { (trackSourceFiles) ->
```

Comment:
Why do we use this / care about the non-trackSourceFiles case for this test (but not the others)?

Reply:
The other tests fail if Anvil isn't tracking the source files.  This one should work in both cases, so the extra test case is just making sure I don't introduce a regression.